File: app/src/main/java/bobzien/com/smartrecyclerviewadapter/ViewHolder/TypedViewHolder2.java
package bobzien.com.smartrecyclerviewadapter.ViewHolder;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.TextView;

import bobzien.com.smartrecyclerviewadapter.R;
import bobzien.com.smartrecyclerviewadapter.SmartRecyclerViewAdapter.TypeViewHolder;
import bobzien.com.smartrecyclerviewadapter.SmartRecyclerViewAdapter.ViewHolder;
import bobzien.com.smartrecyclerviewadapter.SmartRecyclerViewAdapter.Wrapper;
import bobzien.com.smartrecyclerviewadapter.TypedObject;
import butterknife.ButterKnife;
import butterknife.InjectView;

public class TypedViewHolder2 extends TypeViewHolder<TypedObject>
{
    @InjectView(R.id.text)
    protected TextView textView;

    public TypedViewHolder2(@NonNull Context context, @NonNull Class clazz, @Nullable View itemView)
    {
        super(context, clazz, itemView);

        if(itemView != null)
        {
            ButterKnife.inject(this, itemView);
        }
    }

    @Override
    public ViewHolder getInstance(@NonNull View itemView)
    {
        return new TypedViewHolder2(getContext(), getClass(), itemView);
    }

    @Override
    public int getLayoutResourceId()
    {
        return R.layout.list_item_typed2;
    }

    @NonNull
    @Override
    public void bindViewHolder(@NonNull TypedObject item, boolean selected)
    {
        textView.setText("Type " + item.type);
    }

    @Override
    protected boolean canHandle(@NonNull TypedObject item)
    {
        if(item.type == 2)
        {
            return true;
        }
        return false;
    }

    @NonNull
    @Override
    protected Wrapper getWrappedObject(@Nullable final TypedObject item)
    {
        return new Wrapper() {
            @Override
            public Object getItem()
            {
                return item;
            }
        };
    }
}

File: app/src/main/kotlin/bobzien/com/smartrecyclerviewadapter/ViewHolder/TypedViewHolder2.kt
package bobzien.com.smartrecyclerviewadapter.ViewHolder

import android.content.Context
import android.view.View
import android.widget.TextView
import bobzien.com.smartrecyclerviewadapter.R
import bobzien.com.smartrecyclerviewadapter.TypedObject
import bobzien.com.smartrecyclerviewadapter.SmartRecyclerViewAdapter.*

class TypedViewHolder2(context: Context, clazz: Class<TypedObject>, itemView: View?) : TypeViewHolder<TypedObject>(context, clazz as Class<Any>, itemView) {
    protected lateinit  var textView: TextView

    init {
        if (itemView != null) {
            textView = itemView.findViewById(R.id.text) as TextView
        }
    }

    override fun getInstance(itemView: View): ViewHolder<Any> {
        return TypedViewHolder2(context, getHandledClass() as Class<TypedObject>, itemView) as ViewHolder<Any>
    }

    override val layoutResourceId: Int
        get() = R.layout.list_item_typed2

    override fun bindViewHolder(item: TypedObject, selected: Boolean) {
        textView.text = "Type " + item.type
    }

    override fun canHandle(item: TypedObject): Boolean {
        if (item.type == 2) {
            return true
        }
        return false
    }

    override fun getWrappedObject(item: TypedObject?): Wrapper {
        return object : Wrapper {
            override val item: Any
                get() = item as Any
        }
    }
}
